docs/design.md: add design and architecture doc (#1)
confidential-containers/enclave-cc: docs/design.md

to the workload (your application). Instead it uses generic enclaves to isolate your containers from
the rest of the system. Measurement and attestation of the empty generic enclaves is used in
building trust to deliver the application securely.

## Goals

### Keep the same user experience among Confidential Containers

Several existing LibOS containerization solutions pay more attentions to seamlessly running
application into the enclave in enclave-unaware manner. However, those solutions don't provide the
image creation in enclave-unaware manner. Taking SGX LibOS solutions as examples, the image creator
requires extra steps to wrap LibOS runtimes into container image. This is a big burden for the user
experience when using process-based Confidential Containers. Especially each type of LibOS employs
its unique build technology to create container images.

Theoretically, LibOS runtimes are not parts of container workload, so the wrapping operation should
be removed from image creation flow. Instead, LibOS is purely used as a runtime or loader to host
the sensitive workloads in enclave. This approach targets a usage model where the decision to use

> an enclave is at deployment time. One could for instance grab any image out of a registry and
> target the deployment at an enclave.

Reviewer: we could highlight here that we're targeting a use case where existing unmodified encrypted containers can be supported using process-based isolation:

```suggestion
an enclave is at deployment time. One could for instance grab any unmodified encrypted container image out of a registry and
target the deployment at an enclave in the same way it's done with VM-based isolation.
```

by image creator. Meanwhile, this approach can facilitate the development of standardization of
implementing Enclave-CC, which reuses the methodology of image protections already applied to the
container ecosystem, such as image [encryption](https://github.com/containers/ocicrypt-rs) and
[signing](https://github.com/containers/image/blob/main/docs/containers-signature.5.md), instead
of using LibOSes specific approaches.

![no_wrapping.png](docs/images/no_wrapping.png)

### Align with Confidential Containers principles

Confidential Containers design principle tends to utilize and reuse ecosystem to implement
Confidential Containers. More specifically, Confidential Containers requires:

- Container image is always protected, e.g, signed and/or encrypted.
- Download and unpack container image to the secure and attested enclave for the actual workloads
to run with the protections of confidentiality and integrity.
- Prove that sensitive workloads are running on a genuine and trusted hardware TEE.
- Follow up trust and threat model defined by Confidential Containers.

Enclave-CC approach would strictly align with above design principles to implement a general

> process-based Confidential Containers co-existing with Pod-based Confidential Containers. It

Reviewer 2: Let's use VM-based for this meaning.
I don't want to limit our thinking on how we address pods using process isolation.

provides a consistent Kubernetes user experience across hardware TEEs, but also compared with the
existing, non confidential Kubernetes usage patterns. The container workload is unmodified (but
encrypted/signed) and identical across hardware TEEs.

![k8s_ux.png](docs/images/k8s_ux.png)

where:
- [Inclavare Containers Project](https://github.com/inclavare-containers/inclavare-containers) is
a container runtime to launch the process-based Confidential Container for Intel SGX.

- LibOS generally refers to the SGX LibOS implementations, such as
[Gramine](https://github.com/gramineproject/gramine) and
[Occlum](https://github.com/occlum/occlum).

- `rune` is the container runtime for spawning and running enclaves in containers.

- `runelet` is the containerized process used to interact with LibOS through common PAL APIs.

## Architecture

![arch.png](docs/images/arch.png)

> enclave-CC introduces a helper enclave called `stub enclave` which is a fixed function enclave. It is launched by `shim-rune` during pod creation in a private method. More specifically, `stub enclave` is deployed in the form of OCI bundle rather than a container image, and thus `shim-rune` can call `rune` to start it. `stub enclave` has the same life cycle as the pod. Its main job is to receive and then load the actual workload's container image into an `app enclave`.

Reviewer: I wonder if this `stub enclave` is the best term to describe this fixed function enclave that runs enclave agent. Would `agent enclave` or `image service enclave` be better name? It made me confused earlier too. The `app enclave` would be better described as `stub enclave + app`. What do you think? 

```suggestion
enclave-CC introduces a helper enclave called `stub enclave` which is a fixed function enclave. It is launched by `shim-rune` during `RunPodSandbox`. More specifically, `stub enclave` is deployed in the form of OCI bundle rather than a container image, and thus `shim-rune` can call `rune` to start it. `stub enclave` has the same life cycle as the pod. Its main job is to pull/decrypt/unpack and then load the actual workload's container image into an `app enclave`.
```

Author: `stub enclave` would be renamed to `agent enclave`. `app enclave` is dedicated to run user workload so it is still called `app enclave`.

To do this securely both the stub and app enclaves establish trust through attestation.

`stub enclave` is responsible for hosting a LibOS loaded from OCI bundle previously mentioned and
`enclave-agent` on top of the LibOS. `enclave-agent` consists of
[image management service](https://github.com/confidential-containers/image-rs) and
[attestation agent](https://github.com/confidential-containers/attestation-agent). They are the
common pieces from Confidential Containers project.

FUSE encryption filesystem acts as a temporary storage, providing the protections for the metadata

> and OCI bundle. There are many [alternatives](https://nuetzlich.net/gocryptfs/comparison/) to use.

Author: No problem. It is a good idea to create such a section to refer to further design docs and makes sure the main design doc can be merged prior to others.

> and OCI bundle. There are many [alternatives](https://nuetzlich.net/gocryptfs/comparison/) to use.

Author: I create an issue to collect the design topics for the sub-systems/modules: https://github.com/confidential-containers/enclave-cc/issues/3

image creation in enclave-unaware manner. Taking SGX LibOS solutions as examples, the image creator
requires extra steps to wrap LibOS runtimes into container image. This is a big burden for the user
experience when using process-based Confidential Containers. Especially each type of LibOS employs
its unique build technology to create container images.

Theoretically, LibOS runtimes are not parts of container workload, so the wrapping operation should
be removed from image creation flow. Instead, LibOS is purely used as a runtime or loader to host
the sensitive workloads in enclave. This approach targets a usage model where the decision to use
an enclave is at deployment time. One could for instance grab any unmodified container image
(optionally an encrypted container) out of a registry and target the deployment at an enclave.

As a result, the non-workload components running inside enclave are not mandatory to be provided
by image creator. Meanwhile, this approach can facilitate the development of standardization of
implementing Enclave-CC, which reuses the methodology of image protections already applied to the

> container ecosystem, such as image [encryption](https://github.com/containers/ocicrypt-rs) and

Author: Thanks! Fixed. @hdxia

[signing](https://github.com/containers/image/blob/main/docs/containers-signature.5.md), instead
of using LibOSes specific approaches.

> ![no_wrapping.png](docs/images/no_wrapping.png)

Reviewer 3: `docs/` in the link is redundant. The image will not be shown.
Same for the other 2 image links.

Author: Fixed. Thanks for addressing it !!

### Align with Confidential Containers principles

> Confidential Containers design principle tends to utilize and reuse ecosystem to implement

Reviewer 4: enclave-cc will utilize the image service and attestation service components which are the same as CC solution.

> Confidential Containers. More specifically, Confidential Containers requires:

Author: So this is why using the word "protected". "protected" can imply `signed and/or encrypted`.

Enclave-CC approach would strictly align with above design principles to implement a general
process-based Confidential Containers co-existing with VM-based Confidential Containers. It
provides a consistent Kubernetes user experience across hardware TEEs, but also compared with the
existing, non confidential Kubernetes usage patterns. The container workload is unmodified (but
encrypted/signed) and identical across hardware TEEs.

![k8s_ux.png](docs/images/k8s_ux.png)

where:
- [Inclavare Containers Project](https://github.com/inclavare-containers/inclavare-containers) is
a container runtime to launch the process-based Confidential Container for Intel SGX.

- LibOS generally refers to the SGX LibOS implementations, such as
[Gramine](https://github.com/gramineproject/gramine) and
[Occlum](https://github.com/occlum/occlum).

- `rune` is the container runtime for spawning and running enclaves in containers.

- `runelet` is the containerized process used to interact with LibOS through common PAL APIs.

## Architecture

![arch.png](docs/images/arch.png)

enclave-CC introduces a helper enclave called `agent enclave` which is a fixed function enclave.

> It is launched by `shim-rune` during pod creation in a private method. More specifically,

Author: rune and runelet already address these, so enclave-cc doesn't consider it too much.

`agent enclave` is deployed in the form of OCI bundle rather than a container image, and thus
`shim-rune` can call `rune` to start it. `agent enclave` has the same life cycle as the pod.
Its main job is to receive and then load the actual workload's container image into an
`app enclave`.

To do this securely both the agent and app enclaves establish trust through attestation.

`agent enclave` is responsible for hosting a LibOS loaded from OCI bundle previously mentioned

> and `enclave-agent` on top of the LibOS. `enclave-agent` consists of

Author: In order to be compatible with the process model of Gramine, the two services will be combined together to one process on top of libos.

> and `enclave-agent` on top of the LibOS. `enclave-agent` consists of

Reviewer: I was going to bring up this as an open too. I started thinking yesterday do we want to "combine" them together or just run them as separate ones. AFAICS `attestation-agent` does not necessarily have to be per-pod but it could run on the node. This would require secure gRPC though.

> and `enclave-agent` on top of the LibOS. `enclave-agent` consists of

Author: I think hosting them together would be easy to be managed. `agent enclave` as a whole can be managed and provisioned. If AA is hosted in another enclave, it would increase the communication complexity.

> and `enclave-agent` on top of the LibOS. `enclave-agent` consists of

Reviewer: Both options have pros and cons and taking fresh look for what works best for enclave-cc is the best thing to do. A totally different approach could also be PKCS11 for which there are some building blocks available.

[image management service](https://github.com/confidential-containers/image-rs) and
[attestation agent](https://github.com/confidential-containers/attestation-agent). They are the
common pieces from Confidential Containers project.

FUSE encryption filesystem acts as a temporary storage, providing the protections for the metadata
and OCI bundle. There are many [alternatives](https://nuetzlich.net/gocryptfs/comparison/) to use.

> The LibOS running inside `app enclave` needs to establish an attested channel to retrieve the

Author: agent enclave and app enclave can be different. In the most flexible solution, the type of libos can be selected by the annotation. What exactly is the base image?

decryption key to mount FUSE encryption filesystem. The attested channel can be established in

> several ways according to policy:

Author: This needs to be discussed. The policy is global and should be provisioned in an early stage. This policy file is similar to kata-agent config file in kata-cc. However, it is still unclear how kata-cc deploys its config file.

> several ways according to policy:

Reviewer 5: supporting different options sound fine, but it will add another layer of overhead/latency if the app enclave needs to verified remotely. sgx supports local attestation. the policy and manifest (in case local attestation) need to be build into agent enclave bundle to be verified by the remote KBS/verifier.

> several ways according to policy:

Author: Let's take the manifest for Gramine as example. It is per-enclave, so it would be provisioned remotely for app enclave. More specifically, the provisioning should happen, for example, along with retrieving the FUSE key through local attestation. So the problem is how the manifest for app enclave is transferred to agent enclave in advance.

In addition, the so-called policy is interpreted by LibOS. Currently there is no common definition for policy or manifest or whatever denoting a config among different LibOSes.

# Enclave-CC Design and Architecture

> 

Reviewer 2: Spell out CC for people without the context.
```suggestion
Enclave-based Confidential Containers
```

Author: Done. For all resolved items, I will simply click "Resolve conversation" without further acknowledge.

## Motivation

> There are sort of process-based hardware TEEs, e.g, [Intel SGX](https://www.intel.com/content/www/us/en/developer/tools/software-guard-extensions/overview.html) and [RISC-V Keystone](https://keystone-enclave.org/), which are designed to run the sensitive parts of application in the protected and isolated execution environment in order to securely process confidential data.
> 
> Enclave-CC is introduced to provide a reference design of process-based Confidential Containers for CC community, allowing to run a low complexity and small footprint application in Confidential Containers. In this document, the term "enclave" refers to process-based hardware TEE.
> 
> The approach of Enclave-CC decouples the measurement of the workload from the measurement of the enclave. This is exactly what we've been aiming to do with Kata CC: provision a generic secure environment that is easy to measure and then deploy (app) workloads into it.

Reviewer 2: Alternate write-up for your consideration...
```suggestion
The Confidential Containers project provides a way for cloud native applications to choose
Confidential Compute during deployment without modification to the containers. Confidential Compute
comes in two main flavors: process and virtual machine based isolation. The term enclave is often
associated with process-based isolation.

Enclave-CC enables process-based isolation. Process-based isolation is beneficial by drawing the
isolation boundary exactly around each container process. This reduces the Trusted Computing Base,
or the amount of the system your application depends on for security.

Consistent with the VM-based isolation design, the enclave-cc approach does not measure and attest
to the workload (your application). Instead it uses generic enclaves to isolate your containers from
the rest of the system. Measurement and attestation of the empty generic enclaves is used in
building trust to deliver the application securely.```

- Follow up trust and threat model defined by Confidential Containers.

Enclave-CC approach would strictly align with above design principles to implement a general process-based Confidential Containers co-existing with Pod-based Confidential Containers. It provides a consistent Kubernetes user experience across hardware TEEs, but also compared with the existing, non confidential Kubernetes usage patterns. The container workload is unmodified (but encrypted/signed) and identical across hardware TEEs.

![k8s_ux.png](docs/images/k8s_ux.png)

where:
- [Inclavare Containers Project](https://github.com/inclavare-containers/inclavare-containers) is a container runtime to launch the process-based Confidential Container for Intel SGX.

- LibOS generally refers to the SGX LibOS implementations, such as [Gramine](https://github.com/gramineproject/gramine) and [Occlum](https://github.com/occlum/occlum).

- `rune` is the container runtime for spawning and running enclaves in containers.

- `runelet` is the containerized process used to interact with LibOS through common PAL APIs.

## Architecture

![arch.png](docs/images/arch.png)

> Enclave-CC introduces a helper enclave called `stub enclave` which is a fixed function enclave. It is launched by `shim-rune` during pod creation in a private method. More specifically, `stub enclave` is deployed in the form of OCI bundle rather than a container image, and thus `shim-rune` can call `rune` to consume it. `stub enclave` is kept the same life cycle as pod, waiting to load the image into protected enclave memory.  Be aware of that `stub enclave` must be attested by verifier before launching any `app enclaves`. The attestation to `stub enclave` will be triggered by the decryption of `app image`.

Reviewer 2: ```suggestion
Enclave-CC introduces a helper enclave called `stub enclave` which is a fixed function enclave. It
is launched by `shim-rune` during pod creation in a private method. More specifically, `stub
enclave` is deployed in the form of OCI bundle rather than a container image, and thus `shim-rune`
can call `rune` to start it. `stub enclave` has the same life cycle as the pod. Its main job is to
receive and then load the actual workload's container image into an `app enclave`.  
To do this securely both the stub and app enclaves establish trust through attestation.
```

> `stub enclave` is responsible for hosting a LibOS loaded from OCI bundle previously mentioned and `enclave-agent` on top of the LibOS. `enclave-agent` consists of [image management service](https://github.com/confidential-containers/image-rs) and [attestation agent](https://github.com/confidential-containers/attestation-agent). They are the common pieces from Confidential Containers project.

Reviewer 2: As I understand this better, I think it might be clearer if we just call the stub enclave the enclave-agent. It is a fixed function enclave so it's not really a stub that we are going to load different processes into  - unlike the App Enclave that is meant to house any application.
Also the fact that it is implemented using a libos is, I think, kind of ancillary. Like it's good to show the libos, but the point is not so much that this enclave is hosting a libos, as much as that it is a well defined agent that launches the actual workload.

Author: Good points. Let me add some clarifications here.

`stub enclave` is an anchor used in enclave-cc for pod-level representation, taking the similar work as image-rs + attestation-agent. But as you said, the actual entity is `enclave-agent`. Its name shows the real purpose. So calling it `agent enclave` instead of `stub enclave` in the context requiring emphasizing enclave itself is more clearly. Is it?

Reviewer 2: Yes I think it is more clearly called `agent enclave`. I think this would be analogous to the kata-agent in the VM-based design of CC.
If you do not already have more details on this enclave's design, I would be happy to create some as an additional PR.

Reviewer 4: hi Dan, glad to hear from you about the agent enclave's design idea. We had an initial plan on how to Poc the agent enclave. We can make further communication about it.
